valentyusb/usbcore/rx/pipeline.py

```diff
@@ -53,38 +53,25 @@ def __init__(self):
         # Cross the data from the 48MHz domain to the 12MHz domain
         bit_dat = Signal()
         bit_se0 = Signal()
+        bit_valid = Signal()
 #        self.comb += [
 #            bit_dat.eq(nrzi.o_data),
 #            bit_se0.eq(nrzi.o_se0),
 #        ]
 
-        fifo_idle = Signal(4)
-        fifo_delay = Signal(4)
-        fifo_filling = Signal()
-
-        self.sync.usb_48 += If(nrzi.o_valid,
-            If(nrzi.o_data, If(~fifo_idle[3], fifo_idle.eq(fifo_idle + 1))).Else(fifo_idle.eq(0))
-        )
-
         fifo = genlib.fifo.AsyncFIFO(2, 8)
         self.submodules.fifo = fifo = ClockDomainsRenamer({"write":"usb_48", "read":"usb_12"})(fifo)
 
         self.comb += [
             fifo.din[0].eq(nrzi.o_data),
             fifo.din[1].eq(nrzi.o_se0),
-            fifo.we.eq(nrzi.o_valid & (~fifo_idle[3] | ~nrzi.o_data)),
```

Comment:
Mmm, I don't see anything in your changes that would allow to remove this.
The point here is to only send data when there is actual data / activity (i.e. we're in a packet).

Without this, the fifo can just fill up indefinitely and never empty. Because :
 - When there is no packet/data, the clock recovery will just generate 1 strobe every 4 clock (@48M)
 - The read side will pull one bit every cycle at 12 MHz.

So far so good ... and I'm sure in simulation it works.

But when you're inside a packet, if the host clock is ever so slightly faster, the clock recovery will push a bit more than 1 clock every 4 cycles ... and those "extra bits" will never have a chance to empty and so the fifo will just progressively fill up.